projects/driver_controls/src/pedal_calibration.c

```diff
@@ -0,0 +1,192 @@
+                             PedalCalibrationChannel calib_channel,
+                             ThrottleChannel throttle_channel) {
+  PedalCalibrationRange *range = storage->band[calib_channel];
+  storage->settings.throttle_calibration_data->line[throttle_channel] = (ThrottleLine){
+    .full_brake_reading = (range[PEDAL_CALIBRATION_STATE_FULL_BRAKE].max +
+                           range[PEDAL_CALIBRATION_STATE_FULL_BRAKE].min) /
+                          2,
+    .full_throttle_reading = (range[PEDAL_CALIBRATION_STATE_FULL_THROTTLE].max +
+                              range[PEDAL_CALIBRATION_STATE_FULL_THROTTLE].min) /
+                             2,
+  };
+}
+
+static int16_t prv_calc_range(PedalCalibrationRange *range) {
+  return range[PEDAL_CALIBRATION_STATE_FULL_THROTTLE].max -
+         range[PEDAL_CALIBRATION_STATE_FULL_BRAKE].min;
+}
+
+// Based on the points initialized in the bands, it calculates and initializes the data in
+// ThrottleCalibrationData.
+StatusCode pedal_calibration_calculate(PedalCalibrationStorage *storage) {
+  if (storage == NULL) {
+    return status_code(STATUS_CODE_INVALID_ARGS);
+  }
+
+  ThrottleCalibrationData *throttle_calibration = storage->settings.throttle_calibration_data;
+
+  // Compare the range of bands from channels and determine the channels as main or secondary.
+  int16_t range_a = prv_calc_range(storage->band[PEDAL_CALIBRATION_CHANNEL_A]);
+  int16_t range_b = prv_calc_range(storage->band[PEDAL_CALIBRATION_CHANNEL_B]);
+
+  PedalCalibrationChannel main_channel = PEDAL_CALIBRATION_CHANNEL_A;
+  PedalCalibrationChannel secondary_channel = PEDAL_CALIBRATION_CHANNEL_B;
+
+  if (range_a < range_b) {
+    main_channel = PEDAL_CALIBRATION_CHANNEL_B;
+    secondary_channel = PEDAL_CALIBRATION_CHANNEL_A;
+  }
+
+  throttle_calibration->channel_main = storage->settings.adc_channel[main_channel];
+  throttle_calibration->channel_secondary = storage->settings.adc_channel[secondary_channel];
+
+  // Get the main band's vertices.
+  prv_calc_midline(storage, main_channel, THROTTLE_CHANNEL_MAIN);
+  // Get the secondary band's vertices.
+  prv_calc_midline(storage, secondary_channel, THROTTLE_CHANNEL_SECONDARY);
+
+  // Get the full range for main channel so we can calculate the size of each zone.
+  int16_t range = prv_calc_range(storage->band[main_channel]);
+
+  // Use the percentages to calculate the thresholds of each zone.
+  ThrottleZoneThreshold *brake_threshold =
+                            &throttle_calibration->zone_thresholds_main[THROTTLE_ZONE_BRAKE],
+                        *coast_threshold =
+                            &throttle_calibration->zone_thresholds_main[THROTTLE_ZONE_COAST],
+                        *accel_threshold =
+                            &throttle_calibration->zone_thresholds_main[THROTTLE_ZONE_ACCEL];
+  PedalCalibrationRange *calib_range = storage->band[main_channel];
+
+  brake_threshold->min = calib_range[PEDAL_CALIBRATION_STATE_FULL_BRAKE].min;
```

Comment:
There should probably be some assertions here regarding the percentage and tolerances? As it is not it is technically possible to calibrate over 100% with bad inputs unless I missed something elsewhere? I assume these values are hardcoded but static_asserts or something would just avoid messing up.

Reply:
yep, that's the TODO for validating settings